lib/logical_solver/presolve.cpp

```diff
@@ -0,0 +1,889 @@
+        return false;
+      }
+    }
+    // If we have one row of each sign, we now have to check feasibility
+    // for inequalities.
+    if (large_to_small_ratio < 0) {
+      // large bound by ratio (rounded down to ensure feasibility for
+      // integers if solving MIP) now becomes an upper bound of the 
+      // small inequality. Thus, if it is smaller than the small row's 
+      // bound, the problem is infeasible.
+      if (solve_MIP_) {
+        if (std::floor(large_lower_bound_by_ratio) < lower_bounds_.at(small_row_index)) {
+          return false;
+        }
+      } else {
+        if (large_lower_bound_by_ratio < lower_bounds_.at(small_row_index)) {
+          return false;
+        }
+      }
+    }
+    return true;
+
+  };
+
+  void Presolve::updateStateParallelRow(
+    int small_row_index, 
+    int large_row_index,
+    double large_to_small_ratio,
+    double large_lower_bound_by_ratio
+  ) {
+    // turn off large row and log into stack
+    presolve_active_rows_.at(large_row_index) = false;
+    presolve_active_rows_count_ -= 1;
+    struct presolve_log log = {large_row_index, -1, 3};
+    presolve_stack_.push(log);
+
+    // if we have an inequality, update the bound on the small row 
+    // to ensure that both are satisfied.
+    if (small_row_index < inequalities_count_) {
+      // If we are solving the MIP then we have that we 
+      // need to round up the large row's lower bound divided by 
+      // the ratio to ensure that constraints are satisfied when 
+      // restricted to integers.
+      double potential_lower_bound;
+      if (solve_MIP_) {potential_lower_bound = std::ceil(large_lower_bound_by_ratio);}
+      else {potential_lower_bound = large_lower_bound_by_ratio;}
+
+      int small_row_lower_bound = lower_bounds_.at(small_row_index);
+
+      if (large_to_small_ratio > 0) {
+        // If the inequalities both have the same sign then we use 
+        // the bound that takes up all the slack.
+        if (small_row_lower_bound < potential_lower_bound) {
+          lower_bounds_.at(small_row_index) = potential_lower_bound;
+        }
+      } else {
+        // If they do not have the same sign, then that means that the 
+        // lower bound of the large one divided by the large 
+        // to small ratio (rounded down if we are solving the MIP), 
+        // will have become an upper bound of the small one - if 
+        // this upper bound is smaller than the small one's
+        // lower bound, we will have already deemed the system unfeasible in 
+        // checkAreParallelRowsFeasible.
+        if (solve_MIP_) {
+          upper_bounds_.at(small_row_index) = std::floor(large_lower_bound_by_ratio);
+        } else {
+          upper_bounds_.at(small_row_index) = large_lower_bound_by_ratio;
+        }
+      }
+    }
+  };
+
+  void Presolve::applyParallelRowPostsolve(int row_index) {
+    if (isRowActivePostsolve(row_index)) {
+      if (checkConstraint(row_index, 3)) {
```

Comment:
Not 100% necessary so don't stress about it if you're short on time but where do these numbers come from? If they have a meaning then better practice would probably be to give them a name and store them as a class data member or something?

Reply:
Might do if we have time.

Reply:
They IDs for the rules, might do the class data member at some point.